File: tree/src/main/java/team/dedica/logo/tree/GrowthParameters.java
package team.dedica.logo.tree;

import processing.core.PVector;

public interface GrowthParameters<T> {

    PlantSegment<T> getSeed(PVector origin, float scaleRatio);

    boolean canStillGrow(PlantSegment<T> plant);

    float getVelocityReduction();

    /**
     * slight bump into other direction
     * @return bump vector
     */
    PVector bump();

    float getRandomVelocityFactor(PlantSegment<T> plant);

    boolean isWithinBounds(PVector location);

    float getBranchProbabilityFactor();

    PVector createGrowthVector(PlantSegment<T> plant);

    float getNewDiameterFactor();

    /**
     * @param segment given segment
     * @return true if it can have child branches
     */
    boolean canBranch(PlantSegment<T> segment);
}

File: src/main/java/team/dedica/logo/forest/items/GrowthParameters.kt
package team.dedica.logo.forest.items

import processing.core.PVector

interface GrowthParameters {

    fun getVelocityReduction(): Float
    fun getBranchProbabilityFactor(): Float
    fun getNewDiameterFactor(): Float
    fun getSeed(origin: PVector, scaleRatio: Float): PlantSegment
    fun canStillGrow(plant: PlantSegment): Boolean


    /**
     * slight bump into other direction
     * @return bump vector
     */
    fun bump(): PVector
    fun getRandomVelocityFactor(plant: PlantSegment): Float
    fun isWithinBounds(location: PVector): Boolean

    fun createGrowthVector(plant: PlantSegment): PVector

    /**
     * @param segment given segment
     * @return true if it can have child branches
     */
    fun canBranch(segment: PlantSegment): Boolean
}
